### PR TITLE
Draft: Implement native NLR on Windows under MSVC

### PR DESCRIPTION
**Pull request by KubaO**
Native NLR now works when building with MSVC for both x86 and x64.

This is one step on the way to getting feature parity between MINGW/gcc Windows builds and MSVC.

### CODE REVIEW

**Comment by stinos** [comment]
Out of interest: how did you get the asm code? Is it from the disassembly of the msvc setjmp implementation? It looks rather similar but that's no surprise, after are there's only so many ways to do this. Just for reference it's shown here https://github.com/micropython/micropython/issues/4652). That being said: since setjmp is available and works it's not clear to me what we're gaining by switching to a custom implementation.

File: mpy-cross/mpy-cross.vcxproj [hunk 1]
```diff
@@ -45,6 +45,7 @@
   </PropertyGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.props" />
   <ImportGroup Label="ExtensionSettings">
+    <Import Project="$(VCTargetsPath)\BuildCustomizations\masm.props" />
```
Comment (stinos) [inline mpy-cross/mpy-cross.vcxproj:48]: Can you make indentation the same as the rest?

File: ports/windows/micropython.vcxproj [hunk 2]
```diff
@@ -103,11 +104,18 @@
...
   <Import Project="msvc/genhdr.targets" />
   <Import Project="$(CustomPropsFile)" Condition="exists('$(CustomPropsFile)')" />
   <Target Name="GenerateMicroPythonSources" BeforeTargets="BuildGenerateSources" DependsOnTargets="GenerateHeaders;FreezeModules">
   </Target>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.targets" />
   <ImportGroup Label="ExtensionTargets">
+    <Import Project="$(VCTargetsPath)\BuildCustomizations\masm.targets" />
   </ImportGroup>
-</Project>
+</Project>
```
Comment (stinos) [inline ports/windows/micropython.vcxproj:121]: Newline shouldn't change here

File: py/nlr.c [hunk 3]
```diff
@@ -28,7 +28,9 @@
 
 #if !MICROPY_NLR_SETJMP
```
Comment (stinos) [inline py/nlr.c:29]: This will still be defined for msvc builds so the new code isn't actually used, or am I missing something?

**Comment by KubaO** [comment]
> Out of interest: how did you get the asm code?

I wrote it from first principles, basing on the existing code in mpy but adapting it to MS calling conventions. Never looked at MSVC code. Since custom asm NLR seems to be available and used for most targets on gcc/clang, I thought perhaps it wouldn’t be a bad idea to get it done for MSVC too.

This is a first step toward getting native code emitters - at least for Intel - working on MSVC (already done just needs to be cleaned up). Thinking about it now, perhaps I should bite the bullet and get ARM NLR and native emitter working user MSVC too.

File: ports/windows/micropython.vcxproj (hunk 2 above)
Comment (KubaO) [reply ports/windows/micropython.vcxproj:121]: New line is what MSVC nowadays puts at the end of the file if anyone ever wanted to open it in non-read only mode. I can of course remove it.

File: py/nlr.c (hunk 3 above)
Comment (KubaO) [reply py/nlr.c:29]: It seems not to be defined under MSVC. The new implementations definitely end up being used. I had to debug them :)

**Comment by KubaO** [comment]
Turns out Windows NLR handling will also need SEH state to be preserved.

So this is a draft at the moment. I'll investigate whether the setjmp/longjmp implementation is sufficient.

**Pull request closed without merging**